indico/modules/search/controllers.py

```diff
@@ -5,60 +5,77 @@
 # modify it under the terms of the MIT License; see the
 # LICENSE file for more details.
 
-from flask import session
-from marshmallow import fields
+import math
+
+from flask import request, session
+from marshmallow import INCLUDE, fields
 from sqlalchemy.orm import undefer
+from werkzeug.exceptions import BadRequest
 
 from indico.core.db import db
 from indico.core.db.sqlalchemy.protection import ProtectionMode
 from indico.modules.categories import Category
 from indico.modules.events import Event
 from indico.modules.groups import GroupProxy
 from indico.modules.search.base import IndicoSearchProvider, SearchTarget, get_search_provider
-from indico.modules.search.schemas import CategoryResultSchema, EventResultSchema
+from indico.modules.search.schemas import DetailedCategorySchema, EventSchema
 from indico.modules.search.views import WPSearch
+from indico.util.caching import memoize
 from indico.web.args import use_kwargs
 from indico.web.rh import RH
 
 
+@memoize
+def get_groups(user):
+    access = [user.identifier] + [x.identifier for x in user.local_groups]
+    if user.can_get_all_multipass_groups:
+        access += [GroupProxy(x.name, x.provider.name).identifier
+                   for x in user.iter_all_multipass_groups()]
+    return access
+
+
 class RHSearchDisplay(RH):
     def _process(self):
         return WPSearch.render_template('search.html')
 
 
-# TODO: plugin disabled
 class RHAPISearch(RH):
-    TARGET = None
-    USE_PROVIDER = True
-
-    def search(self, page, q):
-        return {'pages': 0, 'total': 0, 'results': []}
-
     @use_kwargs({
         'page': fields.Int(missing=1),
         'q': fields.String(required=True),
-    }, location='query')
-    def _process(self, page, q):
+    }, location='query', unknown=INCLUDE)
+    def _process(self, page, q, **params):
+        target = SearchTarget.get(request.view_args['type'], SearchTarget.event)
+        if not target:
+            raise BadRequest('Invalid search target')
```

Comment:
Are you sure this will ever happen? You are passing a default value to `.get()`

Reply:
No, it doesn't make sense, I was undecided initially, but I believe my latest idea was to default it to Events.

Reply:
Isn't it coming from `any()` in the url rule? If yes, I'd do `SearchTarget[request.view_args['type']]` since it makes sense to fail loudly if we get crap data there.

But actually, the even nicer option would be this:

```python
@use_kwargs({'type': EnumField(SearchTarget)}, location='view_args')
```